# User increase for apps

We are building apps that are free to download and install, our main source of revenue consists of in-app ads. Our goal for this project is to help our developers understand what type of apps are likely to attract more users.

In [1]:
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apple_data = list(read_file)
apple = apple_data[1:]

def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
        

In [2]:
explore_data(apple_data[1:], 0, 5, rows_and_columns=True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1']


Number of rows: 7197
Number of columns: 16


In [3]:
opened_file = open('googleplaystore.csv')
from csv import reader
read_file = reader(opened_file)
googleplay_data = list(read_file)
googleplay = googleplay_data[1:]

In [4]:
explore_data(googleplay_data[1:],0,5,rows_and_columns=True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up']


Number of rows: 10841
Number of columns: 13


In [5]:
explore_data(apple_data,0,1)
explore_data(googleplay_data,0,1)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']




## relevant values
Apart from the *name* we will have a few relevant values for our project.
* Firstly we are interested in free apps
  We will take the *price* category
* Secondly we would like to see the type of the app:
  In applestore the *genre* is `'prime_genre'`
  In googleplay it is `'genres'`
* Thirdly we will be looking at volume
  This differs within the two datasets. Googleplay has 
  `'installs'` but `'reviews count'` is more useful.
  the applestore only has the *amount of ratings* as
  `'rating_count_tot'`.
  
In order to check out the column names check this [link](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)

Here is an overview of the relevant values.

| value  | apple index  | googleindex  |
|---|---|---|
| name  | 1  | 0  |
| price  | 4  | 7  |
| genre  | 11  | 9  |
| rating count  | 5  | 3  |


In [6]:
explore_data(googleplay_data[1:],10472,10473)

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']




In [7]:
for app in googleplay:
    name = app[0]
    if name == 'Instagram':
        print(app)


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


In [8]:
duplicate_apps = []
unique_apps = []

for app in googleplay:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps))
print('\n')
print('Examples of duplicate apps:', duplicate_apps[:15])

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


## Duplicates
We have seen that in googleplay there are 1181 duplicates. Some of the apps are even in there 4 times like 'Instagram' above. 

I would like to have the most recent entry in the dataset, hence I will remove the duplicates with the least reviews. 

Let us see if this is also the case for apple.

In [11]:
duplicate_apps_apple = []
unique_apps_apple = []

for app in apple:
    name = app[0]
    if name in unique_apps:
        duplicate_apps_apple.append(name)
    else:
        unique_apps_apple.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps_apple))
print('\n')
print('Examples of duplicate apps:', duplicate_apps_apple[:15])

Number of duplicate apps: 0


Examples of duplicate apps: []


### We're lucky
no duplicates

## Cleaning the duplicates
In the first code below we can see a dictionary where every key is linked to the highest review score. Which is the most recent value for the app. 
If the name is in the dictionary a second time we have made the code update the value in the dictionary, so that there can only be one entry of every name in the dictionary. 

Afterwards we will use this dictionary to create two lists. One where the rows correspond to our new dictionary, this will be our *clean dataset* 

The rest will be added to another list.

In the end we have explored the clean list in order to check if the number is corresponding to the number of the dictionary.


In [13]:
reviews_max = {}

for app in googleplay:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

In [14]:
print('Expected length:', len(googleplay) - 1181)
print('Actual length:', len(reviews_max))

Expected length: 9659
Actual length: 9659


In [15]:
googleplay_clean = []
already_added = []

for app in googleplay:
    name = app[0]
    n_reviews = float(app[3])
    if (n_reviews == reviews_max[name]) and (name not in already_added):
        googleplay_clean.append(app)
        already_added.append(name)

explore_data(googleplay_clean,0,1,rows_and_columns = True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


Number of rows: 9659
Number of columns: 13


In [16]:
def englishcheck(string):
    non_ascii = 0
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
            if non_ascii > 3:
                return False
            
    return True

print(englishcheck('Instagram'))
print(englishcheck('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(englishcheck('Docs To Go™ Free Office Suite'))
print(englishcheck('Instachat 😜'))

True
False
True
True


# Englishcheck
We are only interested in building English apps, therefore any other data is irrelevant to our inquiry. 
To find out if an app is english or not we are looking for characters outside of the ASCII range of 127. The normal range for English. 

We would also like to make the function able to detect english with a few characters outside of the 127 range. In case of an emoji or otherwise. 

Therefore we have decided to allow up to three characters outside of the 127 range to determine something to be still english. 

In [17]:
apple_english = []
for row in apple:
    name = row[1]
    if englishcheck(name) is True:
        apple_english.append(row)
        
explore_data(apple_english,0,1,rows_and_columns = True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


Number of rows: 6183
Number of columns: 16


In [18]:
googleplay_english = []
for row in googleplay_clean:
    name = row[0]
    if englishcheck(name) is True:
        googleplay_english.append(row)
        
explore_data(googleplay_english,0,3,rows_and_columns = True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


In [19]:
apple_final = []
apple_english_paid = []

for row in apple_english:
    price = float(row[4])
    if price == 0.0:
        apple_final.append(row)
    else:
        apple_english_paid.append(row)
        
print(explore_data(apple_final,0,3,True))

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 3222
Number of columns: 16
None


In [20]:
googleplay_final = []

for row in googleplay_english:
    price = row[7]
    if price == '0':
        googleplay_final.append(row)
        
print(explore_data(googleplay_final,0,3,True))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 8864
Number of columns: 13
None


In our final datasets which have removed the apps which are not free we have 8864 apps in google play and 3222 apps in the apple store. 

###our aim
Our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful on both markets.

For this we will use the 'prime_genre' column in apple:

* Apple_final: column 11

For googleplay we would like to inspect 'Genres' and 'Category'.

* Category: column 1
* Genres: column 9

In [24]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage 
    
    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [25]:
display_table(apple_final,11)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


## App Store

As we can see from free English apps there is a clear winner for genres. Namely *games* with an impressive 58% of the total. The runner up is *entertainment* with 'only' 8%. 

As a general impression we are able to see that there are more apps designed for entertainment (games, photo and video, social networking, sports, music) than for practical purposes (education, shopping, utilities, productivity, lifestyle). 

Unfortunately we are unable to see the amount of users for each app. It is imaginable that *games* could still have less users than *entertainment*. On the other hand, it is probable that designing *games* apps is profitable in general because many of them have stood the test of time.

In [26]:
display_table(googleplay_final,1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

In [27]:
display_table(googleplay_final,9)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

## Googleplay 

### category
In contrast to the English language and free AppStore, Games is not the biggest category falling a little short of 10%. Family is the most frequent with almost 19%. Third and fourth are Tools with 8.5% and Business with 4%. 

After a quick peek on google play, we can see that the family category are mostly games. So we see again that games are the most common category, but not as dominant as in the AppStore.

### Genres
In Genres the results are not very clear. It seems as if entertainment and games are almost gome. But then we can see that there are a lot of subgenres for that. This makes it hard to draw conclusions for this table. 

At least we can say again that there are a lot of subgenres for games. Therefore we will use the categories data to combine with the numbers.

In [32]:
genre_apple = freq_table(apple_final,11)

for genre in genre_apple:
    total = 0
    len_genre = 0
    for app in apple_final:
        genre_app = app[11]
        if genre_app == genre:
            user_ratings = float(app[5])
            total += user_ratings
            len_genre += 1
    avg_user_ratings = total / len_genre
    print(genre, ':', avg_user_ratings)

Lifestyle : 16485.764705882353
Navigation : 86090.33333333333
Education : 7003.983050847458
Book : 39758.5
Social Networking : 71548.34905660378
Medical : 612.0
Sports : 23008.898550724636
Business : 7491.117647058823
Photo & Video : 28441.54375
Entertainment : 14029.830708661417
Shopping : 26919.690476190477
Music : 57326.530303030304
Travel : 28243.8
Utilities : 18684.456790123455
Productivity : 21028.410714285714
Health & Fitness : 23298.015384615384
Food & Drink : 33333.92307692308
Catalogs : 4004.0
Games : 22788.6696905016
Finance : 31467.944444444445
News : 21248.023255813954
Reference : 74942.11111111111
Weather : 52279.892857142855


### analysis genre average users

We have calculated the average number of reviews per app genre in the AppStore. There are a few genres here that draw our attention. 

The first is 'Navigation' with 86000 average user ratings. Second is 'Reference' with an average of around 75000 ratings. And 'Social networking' with 71500 ratings. 

When one considers the amount of work to be done for these categories we would recommend to go for the 'Reference' category. That is a category filled with books and other references like programming languages. As it seems there is a great demand for those apps and the content is more easy to create then a succesful 'Social network' or 'Navigation' app

'Photo & Video' 28441 is still listed in the top10 but I believe it is a very diverse category. There are many users and there is a limitless amount of new stickers and skins to create.

In [39]:
category_table = freq_table(googleplay_final, 1)

for category in category_table:
    total = 0 
    len_category= 0
    for row in googleplay_final:
        category_app = row[1]
        if category_app == category:
            n_installs = row[5]
            n_installs = n_installs.replace(',','')
            n_installs = n_installs.replace('+','')
            total += float(n_installs)
            len_category += 1
    
    avg_n_installs = (total / len_category)
    print(category, ':', avg_n_installs)
        

TOOLS : 10801391.298666667
MEDICAL : 120550.61980830671
BEAUTY : 513151.88679245283
LIFESTYLE : 1437816.2687861272
WEATHER : 5074486.197183099
BOOKS_AND_REFERENCE : 8767811.894736841
TRAVEL_AND_LOCAL : 13984077.710144928
PARENTING : 542603.6206896552
SPORTS : 3638640.1428571427
SHOPPING : 7036877.311557789
LIBRARIES_AND_DEMO : 638503.734939759
ENTERTAINMENT : 11640705.88235294
FINANCE : 1387692.475609756
FAMILY : 3695641.8198090694
PHOTOGRAPHY : 17840110.40229885
VIDEO_PLAYERS : 24727872.452830188
PRODUCTIVITY : 16787331.344927534
NEWS_AND_MAGAZINES : 9549178.467741935
BUSINESS : 1712290.1474201474
GAME : 15588015.603248259
SOCIAL : 23253652.127118643
COMICS : 817657.2727272727
HEALTH_AND_FITNESS : 4188821.9853479853
EDUCATION : 1833495.145631068
COMMUNICATION : 38456119.167247385
ART_AND_DESIGN : 1986335.0877192982
HOUSE_AND_HOME : 1331540.5616438356
DATING : 854028.8303030303
MAPS_AND_NAVIGATION : 4056941.7741935486
FOOD_AND_DRINK : 1924897.7363636363
PERSONALIZATION : 5201482.612244

### Googleplay analysis

If we compare googleplay with the appstore we will need to see which one is more important to focus on.

[link](https://www.counterpointresearch.com/global-smartphone-share/) shows the image from 2018 - 2020. This shows Apple only has 20% of the marketshare left. Android phones like Samsung and Huawei are more popular and cheaper. This leads to the conclusion that it would be better to start out our free app in the Google Play store, and as a result take its data a little more seriously than the AppStore's

Of course this is no reason to ignore the results from the AppleStore, however I would take Google play more seriously concerning the amount users in the future.

Let us sum up the highest scoring categories here.
COMMUNICATION : 38.456.119
VIDEO_PLAYERS : 24.727.872
SOCIAL : 23.253.652
PHOTOGRAPHY : 17.840.110
PRODUCTIVITY : 16.787.331
GAME : 15.588.015

BOOKS_AND_REFERENCE : 8.767.811

Similar to the appstore Video players seems to be our category to pay attention to. The best way to have a great scoring app is to be able to edit video e.g. with stickers and skins. 
